src/derivations/Segmentation.js

```diff
@@ -0,0 +1,278 @@
+        const dataset = this.dataset;
+        dataset.NumberOfFrames = NumberOfFrames;
+        dataset.PixelData = new ArrayBuffer(
+            (dataset.Rows * dataset.Columns * NumberOfFrames) / 8
+        );
+    }
+
+    /**
+     * addSegment - Adds a segment to the dataset.
+     *
+     * @param  {type} Segment   The segment metadata.
+     * @param  {Uint8Array} pixelData The pixelData array containing all
+     *                          frames of segmentation.
+     * @param  {Number[]} InStackPositionNumbers  The frames that the
+     *                                            segmentation references.
+     * @param  {Boolean} [isBitPacked = false]    Whether the suplied pixelData
+     *                                            is already bitPacked.
+     *
+     */
+    addSegment(
+        Segment,
+        pixelData,
+        InStackPositionNumbers,
+        isBitPacked = false
+    ) {
+        if (this.dataset.NumberOfFrames === 0) {
+            throw new Error(
+                "Must set the total number of frames via setNumberOfFrames() before adding segments to the segmentation."
+            );
+        }
+
+        let bitPackedPixelData;
+
+        if (isBitPacked) {
+            bitPackedPixelData = pixelData;
+        } else {
+            bitPackedPixelData = BitArray.pack(pixelData);
+        }
+
+        this._addSegmentPixelData(bitPackedPixelData, isBitPacked);
+        const ReferencedSegmentNumber = this._addSegmentMetadata(Segment);
+        this._addPerFrameFunctionalGroups(
+            ReferencedSegmentNumber,
+            InStackPositionNumbers
+        );
+    }
+
+    _addSegmentPixelData(bitPackedPixelData) {
+        const dataset = this.dataset;
+
+        const pixelDataUint8View = new Uint8Array(dataset.PixelData);
+        const existingFrames = dataset.PerFrameFunctionalGroupsSequence.length;
+        const offset = (existingFrames * dataset.Rows * dataset.Columns) / 8;
+
+        for (let i = 0; i < bitPackedPixelData.length; i++) {
+            pixelDataUint8View[offset + i] = bitPackedPixelData[i];
+        }
+    }
+
+    _addPerFrameFunctionalGroups(
+        ReferencedSegmentNumber,
+        InStackPositionNumbers
+    ) {
+        const PerFrameFunctionalGroupsSequence = this.dataset
+            .PerFrameFunctionalGroupsSequence;
+
+        const isMultiframe = Normalizer.isMultiframeDataset(
+            this.referencedDataset
+        );
+
+        for (let i = 0; i < InStackPositionNumbers.length; i++) {
+            const frameNumber = InStackPositionNumbers[i];
+
+            const perFrameFunctionalGroups = {};
+
+            perFrameFunctionalGroups.PlanePositionSequence = DerivedDataset.copyDataset(
+                this.referencedDataset.PerFrameFunctionalGroupsSequence[
+                    frameNumber - 1
+                ].PlanePositionSequence
+            );
+
+            // TODO -> I think this is write, have someone check in review.
```

Comment:
Can you explain? I am not sure I understand. Are you saying you are not sure what should be those index values?

Reply:
Sorry this should be "I think this is right, have someone check in review." Yeah I think I understood what these values should be, but would like someone more versed to check I'm doing the right thing.